File: fit.pl
#! /usr/local/bin/perl
#
# Usage: fit dir size
#
use warnings;
use strict;
use File::Find;

my $binsize;
my @files;
my @bins;

# -----------------------------------------------------------------------------
# Convert a number to human readable
#
sub num_to_human {
  my $num = shift;
  my $suffix = "B";

  if ($num > 1024 * 1024) {
    $suffix = "M";
    $num /= 1024 * 1024;
  } elsif ($num > 1024) {
    $suffix = "K";
    $num /= 1024;
  }

  if ($suffix eq "B") { return sprintf("%dB", $num); }
  return sprintf("%.2f$suffix", $num);
}

# -----------------------------------------------------------------------------
# Add a name/size tuple to bin bin
#
sub add {
  my ($bin, $name, $size) = @_;
  if ($size > $bin->{free}) {
    printf("Can't fit %s (%s) into %s\n", $name, num_to_human($size),
      num_to_human($bin->{free}));
    exit(1);
  }
  push(@{$bin->{contents}}, [$name, $size]);
  $bin->{free} -= $size;
  $bin->{items}++;
}

# -----------------------------------------------------------------------------
# Get a bin which fits size or create a new one for it
#
sub getbin {
  my $size = shift;
  foreach my $bin (@bins) {
    if ($size <= $bin->{free}) {
        return $bin;
    }
  }
  push(@bins, { "items"    => 0,
                "free"     => $binsize,
                "contents" => [] });
  return $bins[$#bins];
}

# -----------------------------------------------------------------------------
# Get a list of filenames/sizes from path path
#
sub getfiles {
  my $path = shift;
  find(sub{if (-f) {push(@files, [$File::Find::name, -s]);}}, $path);
  # sort by size so large items come first (pop will take largest first)
  @files = sort { $a->[1] - $b->[1] } @files; 
}

# -----------------------------------------------------------------------------
# Main
#
# - Check arguments
# - Initialise the binsize
# - Find all needed files
# - Fit them into bins
# - Print the results
#
if ($#ARGV != 1) {
  print("usage: fit.pl path size\n");
  exit(1);
}

$binsize = $ARGV[1];
if ($binsize =~ /[gG]$/) {
  $binsize =~ s/[^0-9]//g; 
  $binsize *= 1024 * 1024 * 1024;
} elsif ($binsize =~ /[mM]$/) {
  $binsize =~ s/[^0-9]//g;
  $binsize *= 1024 * 1024;
} elsif ($binsize =~ /[kK]$/) {
  $binsize =~ s/[^0-9]//g;
  $binsize *= 1024;
} else {
  $binsize =~ s/[^0-9]//g;
}

getfiles($ARGV[0]);
while ($#files >= 0) {
  my ($name, $size) = @{pop(@files)};
  add(getbin($size), $name, $size);
}

my $total_wasted = 0;                   
foreach my $bin (@bins) {
  foreach my $tuple (@{$bin->{contents}}) {
    my ($name, $size) = @{$tuple};
    print $name."\n";
  }
  printf("==> %.1f%% wasted (%s)\n\n", $bin->{free}/$binsize * 100.0,
    num_to_human($bin->{free}));
  $total_wasted += $bin->{free};
}

printf("Total %d bins, %.1f%% (%s) wasted\n", $#bins+1,
  $total_wasted/(($#bins+1)*$binsize)*100.0, num_to_human($total_wasted));

File: fit.py
#! /usr/local/bin/python -uOO
#
# Usage: fit dir size
#
from __future__ import print_function
import os, sys

def num_to_human(num):
  suffix = 'B'
  if num > 1024*1024:
    suffix = 'M'
    num = num / 1024.0 / 1024.0
  elif num > 1024:
    suffix = 'K'
    num = num / 1024.0
  if suffix == 'B':
    return "{:d}{:s}".format(int(num), suffix)
  return "{:.2f}{:s}".format(num, suffix)

class Bin:
  "A bin to collect files"

  def __init__(self, size):
    self.free     = size
    self.items    = 0
    self.contents = []

  def add(self, name, size):
    if size > self.free:
      raise ValueError("Can't fit {}({}) into {}".format(name, \
        num_to_human(size), num_to_human(self.free)))
    self.contents.append((name, size))
    self.free  = self.free - size
    self.items = self.items + 1


class Fit:
  "Fit a set of files onto a disk"

  def __init__(self, Path, Binsize):
    self.binsize = Binsize 
    self.nbins   = 0
    self.bins    = []
    self.files   = []

    self.getfiles(Path)
    self.fit()
    self.printbins()


  def getfiles(self, path):
    "Read all files and check if they are not too big"
    for root, _, filenames in os.walk(path):
      for filename in filenames:
        fullname = os.path.join(root, filename)
        size = os.path.getsize(fullname)
        if size > self.binsize:
          raise ValueError("Can't fit {}({}) into {}".format(fullname, \
            num_to_human(size), num_to_human(self.binsize)))
        self.files.append((fullname, size))
    # sort by size so large items come first (pop will take largest first)
    self.files.sort(cmp=lambda a, b: a[1] - b[1])

  def getbin(self, size):
    "Get a bin which can hold size or create a new one"
    bestbin = None
    bestfree = self.binsize + 1
    for bin in self.bins:
      if size <= bin.free and bin.free < bestfree:
        bestbin = bin
        bestfree = bin.free
    if bestbin:
      return bestbin

    # no fit, create a new bin
    self.bins.append(Bin(self.binsize))
    self.nbins = self.nbins + 1
    return self.bins[-1]

  def fit(self):
    "Fit all files into bins"
    while self.files:
      name, size = self.files.pop()
      self.getbin(size).add(name, size)

  def printbins(self):
    wasted = 0
    for bin in self.bins:
      for name, _ in bin.contents:
        print(name)
      print("==> {:.1f}% wasted ({:s})".format((float(bin.free) / \
        self.binsize) * 100.0, num_to_human(bin.free)))
      print()
      wasted = wasted + bin.free

    print("Total {} bins, {:.1f}% ({}) wasted".format(self.nbins, \
      float(wasted)/(self.nbins*self.binsize)*100.0, num_to_human(wasted)))

if __name__ == "__main__":
  optlen = len(sys.argv)
  if optlen != 3:
    print("usage: fit dir size_in_mb")
  else:
    sizearg = sys.argv[2]
    size = 0
    if sizearg.endswith('G') or sizearg.endswith('g'):
      size = int(sizearg[:-1]) * 1024 * 1024 * 1024
    elif sizearg.endswith('M') or sizearg.endswith('m'):
      size = int(sizearg[:-1]) * 1024 * 1024
    elif sizearg.endswith('K') or sizearg.endswith('k'):
      size = int(sizearg[:-1]) * 1024
    else:
      size = int(sizearg)

    Fit(sys.argv[1], size)
